Task 4 add student list page

josh-cooper opened the pull request:
## Description
Add a List page to display Student details showing full name with other available attributes.
Add pagination and bootstrap styling to list page.

## Breakdown of Commits
Change students list to display full name : 863be84c99a1a48ed4d59103abde333bb44b2c71

Add pagination to students listing : 349dd0a5a4547203c5a5e487627f376b09487ca3
 - Add will_paginate dependency
 - Change rendering all posts in students index to a paginated query
 - Render page buttons in students index erb
 - Set per page limit to 10

Add bootstrap gem and install static assets : 7b88b6a6920041fb8e7445a1cbb8affe082e66ab

Add bootstrap styling to Students : 05f500ab5658d33214e974ea5643eaa8187cd103
 - Generate default bootstrap layout for Student
 - Replace full name, pagination and show link in generated layout rather than adding in bootstrap styling manually
 - Use rails internationalisation helpers for students table headings

Override will_paginate link renderer : 216f2e040cd5e3a86e8890017a9213c39ad5a78b
    
    Add bootstrap styling to pagination links

Fix decorator interaction with pagination : 956af6b6057f2bbbb3a59d9573269a08169ff282
 - Define PaginatingDecorator to delegate pagination methods to will_paginate

Add custom CSS to Student model show route : 0c06baa6dbbdc5dcca34ad9c7c5302e6872862f9
 - Replace strong tags with spans and custom class
 - Add CSS file asset for default styles for model show routes displaying a list of attributes
  - Right align and bold attribute list headings, left align descriptions
 - Import new CSS asset in application CSS

Add AU I18n : b544ac29f98ce09c63899e275435d9cefcc11c40
 - Add standard AU I18n config file
 - Use long date format in Student model show route

Display flashes : 4f9c2f9e9c0210c5872efd072f6431d5efcf61fd
 - Use bootstrap styling to display flashes
 - Assign bootstrap alert class for particular flash classes
 - Create before_render callback
 - Add error messages to error flashes in before_render callback
 - CSS file for flash styles
 - Remove scaffolded error display

Move students per page default into constant : 39a3030580d87e5999959ca7d9f866b58f4e14e3


Refactor flash_errors to short circuit : 640d900caff9530e476b4857a1e4842cc82e001d

Add specs for error propagation from model to flash : 21acf9bd1752fa314e6f2e13f90c233c4e8a6543

Refactor internal method to access instance variable directly : 19b82cd3dc7de83283b7660a0c09c3b44eb187c8

Refactor erb files to use newer hash syntax : 5554e5f766237ceea0344d79b67ba1becda28022

#### TODO Checklist
* [x] Add a List page to display Student details showing full name with other available attributes.
* [x] Add pagination to the list page with will_paginate gem
* [x] Add css styling with Bootstrap Rails Gem

  
#### Screenshot (If applicable)
![image](https://user-images.githubusercontent.com/41467557/106554670-f3251100-656f-11eb-99c1-365e2a14aaae.png)

josh-cooper committed:
Change students list to display full name

josh-cooper committed:
Add bootstrap gem and install static assets

josh-cooper committed:
Add bootstrap styling to Students

Generate default bootstrap layout for Student
Replace full name, pagination and show link in generated layout rather than adding in bootstrap styling manually
Use rails internationalisation helpers for students table headings

josh-cooper committed:
Override will_paginate link renderer

Add bootstrap styling to pagination links

josh-cooper committed:
Add pagination to students listing

 - Add will_paginate dependency
 - Change rendering all posts in students index to a paginated query
 - Render page buttons in students index erb
 - Set per page limit to 10

josh-cooper committed:
Fix decorator interaction with pagination

Define PaginatingDecorator to delegate pagination methods to will_paginate

jhuizy commented on my-sms/app/views/students/show.html.erb:
```diff
@@ -1,40 +1,31 @@
+  <h1><%=t '.title', :default => model_class.model_name.human.titleize %></h1>
+</div>
+
+<dl class="dl-horizontal">
+  <dt><strong><%= model_class.human_attribute_name(:title) %>:</strong></dt>
+  <dd><%= @student.title %></dd>
+  <dt><strong><%= model_class.human_attribute_name(:first_name) %>:</strong></dt>
+  <dd><%= @student.first_name %></dd>
+  <dt><strong><%= model_class.human_attribute_name(:middle_name) %>:</strong></dt>
+  <dd><%= @student.middle_name %></dd>
+  <dt><strong><%= model_class.human_attribute_name(:last_name) %>:</strong></dt>
+  <dd><%= @student.last_name %></dd>
+  <dt><strong><%= model_class.human_attribute_name(:email) %>:</strong></dt>
+  <dd><%= @student.email %></dd>
+  <dt><strong><%= model_class.human_attribute_name(:birth_date) %>:</strong></dt>
+  <dd><%= @student.birth_date %></dd>
+  <dt><strong><%= model_class.human_attribute_name(:gender) %>:</strong></dt>
+  <dd><%= @student.gender %></dd>
+</dl>
```
I like the use of `dl`, `dt` and `dd` here for semantic html and accessibility. I think the `<strong>'s` could probably be replaced with a `<span class="your-class">` and then some css to style that appropriately (ie. make it bold).

josh-cooper replied:
Added some custom classes in 7d8ac2e028d62a8a9cfb09bbec0907d51adc1c76

Didn't really end up changing much over the default bootstrap CSS applied to dl/dt/dd elements but was a lesson in the asset pipeline basics.

jhuizy commented on my-sms/app/views/students/show.html.erb:
```diff
@@ -1,40 +1,31 @@
+  <dt><strong><%= model_class.human_attribute_name(:email) %>:</strong></dt>
+  <dd><%= @student.email %></dd>
+  <dt><strong><%= model_class.human_attribute_name(:birth_date) %>:</strong></dt>
+  <dd><%= @student.birth_date %></dd>
```
Is the default date format nice and readable?

josh-cooper replied:
Changed the format to something more nice and local via an AU I18n config.
b544ac29f98ce09c63899e275435d9cefcc11c40

jhuizy replied:
Can't seem to see that commit, has it been pushed? (or maybe I'm blind 😄 )

josh-cooper replied:
Sorry I hadn't pushed 😅
Updated now

jhuizy commented on my-sms/app/views/students/_form.html.erb:
```diff
@@ -1,45 +1,76 @@
-        <li><%= msg %></li>
-      <% end %>
-      </ul>
+  <% if @student.errors.any? %>
+    <div id="error_expl" class="panel panel-danger">
+      <div class="panel-heading">
+        <h3 class="panel-title"><%= pluralize(@student.errors.count, "error") %> prohibited this student from being saved:</h3>
+      </div>
+      <div class="panel-body">
+        <ul>
+        <% @student.errors.full_messages.each do |msg| %>
+          <li><%= msg %></li>
+        <% end %>
+        </ul>
+      </div>
     </div>
```
Another way to do this is to put it in a flash and define a way to display the flash on "standard" views/pages, making this error view a bit more re-usable

josh-cooper replied:
Did this in 4f9c2f9e9c0210c5872efd072f6431d5efcf61fd
Defined a way to display flashes on standard views/pages and automatically created error flashes if there were errors on the Student model.

jhuizy commented on my-sms/app/helpers/application_helper.rb:
```diff
@@ -0,0 +1,48 @@
+      tag(:dive, html, container_attributes.merge(class: 'pagination'))
+    end
+  end
+end
```
Did this code block get taken from somewhere? if so could be worth mentioning in a comment

josh-cooper replied:
This is my own, although based it off some examples I saw.

jhuizy reviewed (commented):
Looks pretty good to me, just some minor discussion points

josh-cooper committed:
Add custom CSS to Student model show route

 - Replace strong tags with spans and custom class
 - Add CSS file asset for default styles for model show routes displaying a list of attributes
  - Right align and bold attribute list headings, left align descriptions
 - Import new CSS asset in application CSS

josh-cooper committed:
Add AU I18n

 - Add standard AU I18n config file
 - Use long date format in Student model show route

jhuizy commented on my-sms/app/controllers/students_controller.rb:
```diff
@@ -9,7 +9,9 @@ class StudentsController < ApplicationController
   def index
-    @students = Student.all
+    @students = StudentDecorator.decorate_collection(
+      Student.paginate(page: params[:page], per_page: 10)
```
As mentioned in Matt's PR, consider moving this magic number (`10`) to a constant or appropriately named variable (eg. `DEFAULT_PER_PAGE`)

josh-cooper replied:
Fixed in 39a3030580d87e5999959ca7d9f866b58f4e14e3

josh-cooper committed:
Display flashes

 - Use bootstrap styling to display flashes
 - Assign bootstrap alert class for particular flash classes
 - Create before_render callback
 - Add error messages to error flashes in before_render callback
 - CSS file for flash styles
 - Remove scaffolded error display

josh-cooper committed:
Move students per page default into constant

jhuizy commented on my-sms/app/views/layouts/application.html.erb:
```diff
@@ -1,14 +1,31 @@
+  </head>
+  <body>
+    <!-- display flashes -->
+    <% flash.each do |type, msg| %>
+      <div class="alert <%= get_alert_class(type) %>">
+        <!-- test if message is string or iterable -->
+        <% if msg.respond_to?(:each) %>
+          <ul class="alert-list">
+            <% msg.each do |msg| %>
+              <li class="alert-item"><%= msg %></li>
+            <% end %>
+          </ul>
+        <% else %>
+          <%= msg %>
+        <% end %>
+        <button type="button" class="close" data-dismiss="alert" aria-label="Close">
+          <span aria-hidden="true">&times;</span>
+        </button>
+      </div>
+    <% end %>
```
There was probably a bit of confusion from my last comment, I meant to only show the the fact that there _is_ an error in the flash, and then have more error details displayed inline. Consider replacing this with the `<%= bootstrap_flash %>` I mentioned in slack?

josh-cooper replied:
I got that, just took it as an exercise to dive into what flashes could be used for and investigate the process the rails controller goes through.

I looked more into the bootstrap_flash helper to try and implement it now and it is slightly incompatible. My implementation handles an array of errors and I'd rather keep that implementation than try and put formatting directly into the flash server side.

jhuizy commented on my-sms/app/controllers/students_controller.rb:
```diff
@@ -84,4 +87,10 @@ def destroy
   end
+
+  def flash_errors
+    student.errors.full_messages.each do |error|
+      (flash[:error] ||= []) << error
+    end if student&.errors
```
You can chain the "safe navigation" operator which would remove the need for the trailing `if student&.errors`,

```suggestion
    student&.errors&.full_messages&.each do |error|
      (flash[:error] ||= []) << error
    end
```

I do think the chaining looks a bit ugly though. My personal preference is short circuiting (return early from the method) although this opinion may vary project to project

```ruby
return unless student&.errors
student.errors.full_messages.each do |error|
...
end

josh-cooper replied:
Didn't know about the return unless syntax.
Switched to that in 640d900caff9530e476b4857a1e4842cc82e001d

josh-cooper replied:
Also I added some specs for this section in 21acf9bd1752fa314e6f2e13f90c233c4e8a6543

jhuizy commented on my-sms/app/views/students/index.html.erb:
```diff
@@ -1,35 +1,46 @@
+                      edit_student_path(student), :class => 'btn btn-default btn-xs' %>
+          <%= link_to t('.destroy', :default => t("helpers.links.destroy")),
+                      student_path(student),
+                      :method => :delete,
+                      :data => { :confirm => t('.confirm', :default => t("helpers.links.confirm", :default => 'Are you sure?')) },
+                      :class => 'btn btn-xs btn-danger' %>
```
Minor - but in the plus repo we prefer to use the `{ key: value }` syntax over the `{ :key => value }` syntax for hashes

josh-cooper replied:
Fixed in 5554e5f766237ceea0344d79b67ba1becda28022

jhuizy commented on my-sms/app/controllers/students_controller.rb:
```diff
@@ -84,4 +87,10 @@ def destroy
   end
+
+  def flash_errors
+    student.errors.full_messages.each do |error|
```
should this be `@student` instead?

```suggestion
    @student.errors.full_messages.each do |error|
```

josh-cooper replied:
Fixed in 19b82cd3dc7de83283b7660a0c09c3b44eb187c8

jhuizy commented on my-sms/app/controllers/students_controller.rb:
```diff
@@ -5,11 +5,14 @@ class StudentsController < ApplicationController
   decorates_assigned :student
 
   before_filter :find_student, only: %i[show edit update destroy]
+  before_render :flash_errors, only: %i[edit create update]
```
I'm thinking if you change the `student` to `@student` in the `flash_errors` method you can move this to `before_filter`. Great work on figuring out how to create your own callbacks, but I've not had to use custom callbacks like this before, so I'm a bit suspicious it's required here.

Also another opinion/personal note, I prefer calling the symbols directly rather than use the `%i` macro

```suggestion
  before_filter :flash_errors, only: [:edit, :create, :update]
```
```

josh-cooper replied:
The reason I had to create before_render is because rails does adds the error messages and the errors themselves to the model at different times. The messages are not added in until render is called. Not sure why, maybe because they want to maintain MVC separation?

jhuizy replied:
The errors will be added when the model validations occur. These validations occur when the record is attempted to save to the database (eg. `@student.save`), or when you specifically call the `valid?` method (more info [here](https://guides.rubyonrails.org/active_record_validations.html)). 

Maybe try adding a `return unless @student.invalid?` line to your `flash_errors` method.

josh-cooper replied:
That was a typo sorry 😅. What I meant was the error IDs themselves are added during validation, but corresponding messages for each type of error seem to be only added during render.

jhuizy replied:
Hmm OK, let's leave for now.

Maybe in future, we can stick with original of showing errors above (using what you had before) and just indicate a simple message in the flash to indicate there are errors

something like

```ruby
def render_flash
  if @model.invalid?
    flash[:error] = "There are errors with your model"
  end
end
```

josh-cooper committed:
Refactor flash_errors to short circuit

josh-cooper committed:
Add specs for error propagation from model to flash

josh-cooper committed:
Refactor internal method to access instance variable directly

josh-cooper committed:
Refactor erb files to use newer hash syntax

jhuizy commented on my-sms/spec/controllers/students_controller_spec.rb:
```diff
@@ -125,6 +129,10 @@
+
+      it "should propagate model errors to an error flash" do
+        expect { subject }.to change{ flash[:error] }.from(nil).to(["First name can't be blank"])
+      end
```
nice 👍
